src/formatting/items.rs

```diff
@@ -916,7 +916,7 @@ fn format_impl_ref_and_type(
         result.push_str(format_unsafety(unsafety));
 
         let shape = Shape::indented(offset + last_line_width(&result), context.config);
-        let generics_str = rewrite_generics(context, "impl", generics, shape)?;
+        let generics_str = rewrite_generics(context, "impl", generics, shape, false)?;
```

Comment:
I feel like adding a boolean to the signature for `rewrite_generics` is shifting some of the aspects of formatting generics to the callers :thinking: 

What if instead the new param added to `rewrite_generics` was an [ItemKind](https://docs.rs/rustc-ap-rustc_ast/668.0.0/rustc_ap_rustc_ast/ast/enum.ItemKind.html) and then internally `rewrite_generics` could deal with enabling/disabling the spacing based on the combination of the `ItemKind` and config options.

IIRC not all of these functions that internally call `rewrite_generics` have the original AST Item node, but they could either pass the corresponding ItemKind variant to `rewrite_generics` or could themselves take an extra param